### ISSUE TITLE
Hierarchy of concepts by RF2

### ISSUE THREAD

**Issue by erfelipe** [IHTSDO/snomed-owl-toolkit#36]
Hi @kaicode ,

If you can give me a tip, it would be really useful. I'm trying to extract the hierarchical structure from the RF2 format, when choosing a certain term. The algorithm is working relatively well, with recursive calling and considering a hierarchy of any level.
But I came across an error situation. Even considering the _static_ version.

My example was with the term "heart attack", my code considers the term to be a label and discovers the main concept (22298006) Myocardial infarction (disorder). OK.
Then I take all the axioms related to this concept and make a parse to find out which other concepts are "children" of this main concept.
![image](https://user-images.githubusercontent.com/17271754/72764370-8b02c780-3bc6-11ea-9df2-9a9bb9e67368.png)

However, in this example with 22298006 (Myocardial infarction (disorder)) as the main term, two concepts that appear in the OWL version and on the website (static) according to the images, were not recovered by the axiom. In fact the problem is only in one: 371068009 (**Myocardial infarction with complication (disorder)**), because if he is not considered, your hierarchical child will not be either.
![image](https://user-images.githubusercontent.com/17271754/72764397-ae2d7700-3bc6-11ea-9bcc-b8aee76b4c50.png)

I went looking for the specific axioms of this term and found the following statement:

EquivalentClasses (: **30277009** ObjectIntersectionOf (: 29889000: **371068009**: 57054005 ObjectSomeValuesFrom (: 609 096 000 ObjectIntersectionOf (ObjectSomeValuesFrom (: 116 676 008: 125 671 007) ObjectSomeValuesFrom (: 363 698 007: 21814001))) ObjectSomeValuesFrom (: 609 096 000 ObjectIntersectionOf (ObjectSomeValuesFrom (: 116 676 008: 55470003) ObjectSomeValuesFrom (: 363698007: 74281007)))))

In other words, in the axiom, the concept **371068009** is not linked in **22298006**, but in: **_30277009_**. However, in both Browser and OWL, the structure is linked in 22298006.
I'm looking in the refset file, which has the axioms.

Do you have any suggestions for what can be done, or where am I going wrong?

Once again, thank you for your support.

**Comment by kaicode** [comment]
Hi @erfelipe, 

Axioms of type SubClassOf and EquivalentClasses usually take the form X(A B) where X is the class type, A is a named class (concept in SNOMED) and B is either a named parent class, a list of parent classes or a combination of parents and attributes. A is the subject and B is the stated model with parents and attributes.
(There are some examples where A is an expression rather than a named class. These are General Concept Inclusion axioms (GCIs) and are used by the reasoner to infer more children but I would recommend ignoring these for your use case.)

If we grep the July 2019 RF2 snapshot of axioms filtering by _371068009_ surrounded by tab characters we find the axiom for _371068009 | Myocardial infarction with complication (disorder) |_ and the third column has value `1` which shows that the record is active.
```
grep "\t371068009\t" SnomedCT_InternationalRF2_PRODUCTION_20190731T120000Z/Snapshot/Terminology/sct2_sRefset_OWLExpressionSnapshot_INT_20190731.txt 
478d3004-7abe-4bb3-a2fa-9437c396148c	20190731	1	900000000000207008	733073007	371068009	SubClassOf(:371068009 :22298006)
```
The axiom takes the format X(A B) where X is SubClassOf meaning in SNOMED that the concept is primitive. A is the named class _371068009 | Myocardial infarction with complication (disorder) |_. B is _22298006 | Myocardial infarction (disorder) |_.
This shows that 22298006 is the only stated parent of 371068009.

Try following this method to find the other two stated parents.

Kind regards,
Kai

**Comment by erfelipe** [comment]
@kaicode ,

Thanks again.

My error was to consider the "full terminology" archive:
![image](https://user-images.githubusercontent.com/17271754/72806390-db643e80-3c33-11ea-868a-a563e98bcf70.png)

Best regards ! 
Eduardo Felipe

**Comment by erfelipe** [comment]
Kai,

Perhaps this details can help the SNOMED export code:

1. The sct2_Description_Snapshot-en_INT_20190731.txt have an ["] in one line of file. This " symbol raise an error in code to read the line. 
2. The sct2_StatedRelationship_Snapshot_INT_20190731.txt file have entire lines with 0 (zero) active flag (= inaticve). 

So, my code is working now. Thanks for help. 

Regards,

**Comment by kaicode** [comment]
Hi.
Yes, the snapshot files contain the latest state of all components, many of these are inactive. This is expected.
The description file is tab delimited. There are many lines including quote symbols. This is also expected.
Cheers.

**Comment by erfelipe** [comment]
@kaicode ,

Please, look this situation: 
1. I'm looking for "Heart attack" and I find many ConceptIDs:
![image](https://user-images.githubusercontent.com/17271754/73015225-655e0400-3dfa-11ea-99aa-e4ee96920074.png)

2. Ok. I'm discover that 22298006 is the target conceptID 

3. But, looking in "description" file, I have many descriptions for this ConceptID
![image](https://user-images.githubusercontent.com/17271754/73015513-ef0dd180-3dfa-11ea-8e63-f88e6f0e8613.png)

4. In OWL (non stated) the hierarchy is build with "Myocardial infarction" in the top, the "main description". 
![image](https://user-images.githubusercontent.com/17271754/73015657-34ca9a00-3dfb-11ea-9812-f535e4d25c0c.png)

5. It's possible (without reasoner way) to discover the "main" description in RF2 state files? (I'm try to look some IdKey for each description, but I don't see relationship with another file to do this.) 

Thanks,

**Comment by kaicode** [comment]
To find the main description for a concept select the row from the description file which has active=1 and typeId=900000000000003001. This is the Fully Specified Name of the concept which you can see in Protégé.
